src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -123,17 +123,14 @@ async Task AddDocumentsAndProject(Project? project, ImmutableArray<string> suppo
                     return;
                 }
 
-                var isFSAOn = globalOptions.IsFullSolutionAnalysisEnabled(project.Language);
-                var documents = ImmutableArray<TextDocument>.Empty;
-                // If FSA is on, then add all the documents in the project.  Other analysis scopes are handled by the document pull handler.
-                if (isFSAOn)
-                {
-                    documents = documents.AddRange(project.Documents).AddRange(project.AdditionalDocuments);
-                }
+                // Only process closed files if FSA is on.
+                if (!globalOptions.IsFullSolutionAnalysisEnabled(project.Language))
+                    return;
+
+                var documents = ImmutableArray<TextDocument>.Empty.AddRange(project.Documents).AddRange(project.AdditionalDocuments);
 
-                // If all features are enabled for source generated documents, make sure they are included when FSA is on or a file in the project is open.
-                // This is done because for either scenario we've already run generators, so there shouldn't be much cost in getting the diagnostics.
-                if ((isFSAOn || isOpen) && solution.Services.GetService<IWorkspaceConfigurationService>()?.Options.EnableOpeningSourceGeneratedFiles == true)
+                // If all features are enabled for source generated documents, make sure they are included when computing diagnostics.
+                if (solution.Services.GetService<IWorkspaceConfigurationService>()?.Options.EnableOpeningSourceGeneratedFiles == true)
```

Comment:
processing of closed sg-documents is controlled entirely by if this flag is set or not.